contractsV5/FeeHandler.sol

```diff
@@ -0,0 +1,102 @@
+    uint public totalRewards;
+    mapping(uint => uint) public totalRewardsPerEpoch;
+
+    constructor(IKyberDAO _kyberDAOContract, KyberNetwork _kyberNetworkContract) public {
+        kyberDAOContract = _kyberDAOContract;
+        kyberNetworkContract = _kyberNetworkContract;
+    }
+
+
+    function encodeData(uint _burn, uint _reward, uint _epoch, uint _expiryBlock) public pure returns (uint) {
+        return (((((_burn << BITS_PER_PARAM) + _reward) << BITS_PER_PARAM) + _epoch) << BITS_PER_PARAM) + _expiryBlock;
+    }
+
+    function decodeData() public view returns(uint, uint, uint, uint) {
+        uint expiryBlock = brrAndEpochData & (1 << BITS_PER_PARAM) - 1;
+        uint epoch = (brrAndEpochData / (1 << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        uint rewardInBPS = (brrAndEpochData / (1 << BITS_PER_PARAM << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        uint burnInBPS = (brrAndEpochData / (1 << BITS_PER_PARAM << BITS_PER_PARAM << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        return (burnInBPS, rewardInBPS, epoch, expiryBlock);
+    }
+
+    // Todo: future optimisation to accumulate rebates for 2 rebate wallet
+    // encode totals, 128 bits per reward / rebate
+    function handleFees(address[] calldata eligibleWallets, uint[] calldata rebatePercentages) external payable returns(bool) {
+        // Decoding BRR data
+        (uint burnInBPS, uint rewardInBPS, uint epoch, uint expiryBlock) = decodeData();
+        uint rebateInBPS = BPS - rewardInBPS - burnInBPS;
+
+        // Check current block number
+        if(block.number > expiryBlock) {
+            (burnInBPS, rewardInBPS, rebateInBPS, epoch, expiryBlock) = kyberDAOContract.getLatestBRRData();
+
+            // Update brrAndEpochData
+            brrAndEpochData = encodeData(burnInBPS, rewardInBPS, epoch, expiryBlock);
+        }
+
+        uint rebateWei = rebateInBPS * msg.value / BPS;
+        uint rewardWei = rewardInBPS * msg.value / BPS;
+        for(uint i = 0; i < eligibleWallets.length; i ++) {
+            // Internal accounting for rebates per reserve wallet (totalRebatesPerRebateWallet)
+            totalRebatesPerRebateWallet[eligibleWallets[i]] = rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for total rebates (totalRebates)
+            totalRebates += rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for reward per epoch (totalRewardsPerEpoch)
+            totalRewardsPerEpoch[epoch] += rewardWei;
+            // Internal accounting for total rewards (totalRewards)
+            totalRewards += rewardWei;
+        }
+        // Todo: emit event
+        return true;
+    }
+
+
+    function claimStakerReward(address staker, uint percentageinPrecision, uint epoch) public {
+        // onlyDAO
+        // send reward
+        // update rewardPerEpoch
+        // update totalReward
+    }
+
+    function claimReserveRebate(address reserve) public {
+        // only DAO
```

Comment:
anyone can claim
will send amounts to the rebate wallet address if it has any values saved in array....

please make sure
Always leave 1 wei here.
so we don't zero the value
this to avoid high gas cost of changing value for 0 to non zero

here think on the attack vector....

Reply:
I feel like just to be safe, it is better to have an onlyDAO modifier but at the DAO contract, anyone can call it... 

Noted on your other points.

Reply:
we don't want to add code to the DAO if not necessary.
and if anyone can do it from the DAO
why is it safer?

Reply:
I'm not adding any code to the DAO. I was proposing to add the `onlyDAO` modifier to the FeeHandler to ensure that only the DAO contract can interact with some of these functions but anyone can call the function if they interact with the DAO contract. 

I can't think of a reason why it's safer but just thought that we should limit the possible interactions in order to prevent potential misuse of the function that we didnt think of...